### PR TITLE
Hydra psmac

### PR DESCRIPTION
**Pull request by AndreBiedenkapp**
Added psmac_facade.
Restructured Hydra to use PSMAC for parallel execution

### CODE REVIEW

File: smac/facade/psmac_facade.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+            self.__module__ + "." + self.__class__.__name__)
+
+        self.scenario = scenario
+        self.run_id, self.rng = get_rng(rng, run_id, logger=self.logger)
+        self.kwargs = kwargs
+        self.output_dir = None
+        self.rh = RunHistory(average_cost)
+        self._tae = tae
+        self.tae = tae(ta=self.scenario.ta, run_obj=self.scenario.run_obj)
+        if n_optimizers <= 0:
```
Comment (mlindauer) [inline smac/facade/psmac_facade.py]: Shouldn't it be <= 1?

Comment (AndreBiedenkapp) [reply smac/facade/psmac_facade.py]: Yes thanks for pointing that out.

File: smac/facade/hydra_facade.py [hunk 2]
```diff
@@ -222,37 +177,33 @@ def optimize(self) -> typing.List[Configuration]:
...
+            self.optimizer.output_dir = self.output_dir
+            incs = self.optimizer.optimize()
+            cost_per_conf_v, val_ids, cost_per_conf_e, est_ids = self.optimizer.get_best_incumbents_ids(incs)
+            if self.val_set:
+                to_keep_ids = val_ids[:self.incs_per_round]
+            else:
+                to_keep_ids = est_ids[:self.incs_per_round]
+            config_cost_per_inst = {}
+            incs = incs[to_keep_ids]
+            self.logger.info('Kept incumbents')
```
Comment (KEggensperger) [inline smac/facade/hydra_facade.py]: I do not really understand what this message means. Do you mean "Kept X incumbents" or "Kept incumbents this round" ?
Could you make this message more meaningful or remove it?

File: smac/facade/hydra_facade.py [hunk 3]
```diff
@@ -264,10 +215,11 @@ def optimize(self) -> typing.List[Configuration]:
             self.tae = ExecuteTARunHydra(ta=self.scenario.ta, run_obj=self.scenario.run_obj,
                                          cost_oracle=self.cost_per_inst, tae=self._tae)
 
-            self.scenario.output_dir = os.path.join(self.top_dir, "smac3-output_%s" % (
+            self.scenario.output_dir = os.path.join(self.top_dir, "psmac3-output_%s" % (
                 datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d_%H:%M:%S_%f')))
             self.output_dir = create_output_directory(self.scenario, run_id=self.run_id, logger=self.logger)
-        self.rh.save_json(fn=os.path.join(self.top_dir, 'all_runs_runhistory.json'), save_external=True)
+        read(self.rh, os.path.join(self.top_dir, 'psmac3*', 'run_' + str(MAXINT)), self.scenario.cs, self.logger)
```
Comment (KEggensperger) [inline smac/facade/hydra_facade.py]: Very minor remark: As this method is by now used by two scripts could we maybe move it to utils?

File: smac/facade/hydra_facade.py [hunk 4]
```diff
@@ -293,98 +245,32 @@ def _update_portfolio(self, incs: np.ndarray) -> typing.Union[np.float, float]:
...
-        keep_incumbents = incs[to_keep_ids]
-        for kept in keep_incumbents:
-            self.portfolio.append(kept)
-            cost_per_inst = config_cost_per_inst[kept]
-            if self.cost_per_inst:
-                if len(self.cost_per_inst) != len(cost_per_inst):
-                    raise ValueError('Num validated Instances mismatch!')
-                for key in cost_per_inst:
-                    self.cost_per_inst[key] = min(self.cost_per_inst[key], cost_per_inst[key])
+        if self.val_set:  # we have validated data
```
Comment (KEggensperger) [inline smac/facade/hydra_facade.py]: Shouldn't it rather be "we have validation instances"

File: smac/facade/psmac_facade.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+import copy
+
+import pickle
+import multiprocessing
+
+import numpy as np
+
+from ConfigSpace.configuration_space import Configuration
+
+from smac.tae.execute_ta_run_hydra import ExecuteTARunOld
```
Comment (KEggensperger) [inline smac/facade/psmac_facade.py]: Any specific reason why this is called "run_hydra"?

File: smac/facade/psmac_facade.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    Unbound method to be called in a subprocess
+
+    Parameters
+    ----------
+    queue: multiprocessing.Queue
+        incumbents (Configurations) of each SMAC call will be pushed to this queue
+    scenario: Scenario
+        smac.Scenario to initialize SMAC
+    tae: ExecuteTARun
+        Target Algorithm Runner (supports old and aclib format)
```
Comment (KEggensperger) [inline smac/facade/psmac_facade.py]: "old and [new]"?

File: smac/utils/io/cmd_reader.py [hunk 7]
```diff
@@ -401,18 +401,23 @@ def _add_main_options(self):
                               help="[dev] number of hydra iterations. Only active if mode is set to Hydra")
         req_opts.add_argument("--hydra_validation",
```
Comment (KEggensperger) [inline smac/utils/io/cmd_reader.py]: Why is this a req_opt?

**Pull request merged**